sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTracker.java

```diff
@@ -0,0 +1,177 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms.splittabledofn;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static com.google.common.base.Preconditions.checkNotNull;
+import static com.google.common.base.Preconditions.checkState;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import javax.annotation.Nullable;
+import org.apache.beam.sdk.io.range.ByteKey;
+import org.apache.beam.sdk.io.range.ByteKeyRange;
+import org.apache.beam.sdk.transforms.DoFn;
+
+/**
+ * A {@link RestrictionTracker} for claiming {@link ByteKey}s in a {@link ByteKeyRange} in a
+ * monotonically increasing fashion. The range is a semi-open bounded interval [startKey, endKey)
+ * where the limits are both represented by ByteKey.EMPTY.
+ */
+public class ByteKeyRangeTracker extends RestrictionTracker<ByteKeyRange, ByteKey> {
+  private ByteKeyRange range;
+  @Nullable private ByteKey lastClaimedKey = null;
+  @Nullable private ByteKey lastAttemptedKey = null;
+
+  private ByteKeyRangeTracker(ByteKeyRange range) {
+    this.range = checkNotNull(range);
+  }
+
+  /**
+   * Instantiates a new {@link ByteKeyRangeTracker} with the specified range. The keys in the range
+   * are left padded to be the same length in bytes.
+   */
+  public static ByteKeyRangeTracker of(ByteKeyRange range) {
+    return new ByteKeyRangeTracker(ByteKeyRange.of(range.getStartKey(), range.getEndKey()));
+  }
+
+  @Override
+  public synchronized ByteKeyRange currentRestriction() {
+    return range;
+  }
+
+  @Override
+  public synchronized ByteKeyRange checkpoint() {
+    checkState(lastClaimedKey != null, "Can't checkpoint before any successful claim");
+    final ByteKey nextKey = next(lastClaimedKey);
+    // hack to force range to be bigger than the range upper bundle because ByteKeyRange *start*
```

Comment:
<!--new_thread; commit:d28b9222a4c730ab557091334c555aa8fe89bc67; resolved:0-->
* Do you mean "range upper limit" rather than "range upper bundle"?
* Could you give an example where this case can be hit in practice?
* This comment says that start must be < end, but line 68 says it's <= - which condition is it really? I think it should be <= and degenerate ranges like [a, a) should be allowed, unless there's some pathological situation where we'll suffer significant performance overhead due to abundance of such ranges.

Reply:
> - Do you mean "range upper limit" rather than "range upper bundle"?

Yes, I will fix, good catch, thanks

> - Could you give an example where this case can be hit in practice?

This only happens because of `ByteKeyRange` validation checkArgument validation, see:
https://github.com/apache/beam/blob/229e69d241049719fb53adcacab85292d876d7f9/sdks/java/core/src/main/java/org/apache/beam/sdk/io/range/ByteKeyRange.java#L283-L287

I hit this case while reproducing the unit tests based on `OffsetRangeTracker`, see:
https://github.com/iemejia/beam/blob/d28b9222a4c730ab557091334c555aa8fe89bc67/sdks/java/core/src/test/java/org/apache/beam/sdk/transforms/splittabledofn/ByteKeyRangeTrackerTest.java#L89-L100

> - This comment says that start must be < end, but line 68 says it's <= - which condition is it really? I think it should be <= and degenerate ranges like [a, a) should be allowed, unless there's some pathological situation where we'll suffer significant performance overhead due to abundance of such ranges.

Agree, but remember I am limited by `ByteKeyRange` not allowing degenerate ranges so what do you think, should I change `ByteKeyRange` to allow start == end ?